src/middlewared/middlewared/plugins/asigra.py

```diff
@@ -0,0 +1,358 @@
+            )
+            output = (await proc.communicate())[0].decode()
+            if proc.returncode != 0:
+                self.logger.error(output)
+                self.logger.error('Failed to initialize postgresql:\n{}'.format(output))
+                return False
+
+            self.logger.debug(output)
+
+            await await Popen(
+                ['/usr/sbin/chown', '-R', '{}:{}'.format(self.pg_user, self.pg_group), asigra_postgresql_path],
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+                close_fds=True
+            ).wait()
+
+            if not os.path.exists(asigra_postgresql_conf):
+                self.logger.error("{} doesn't exist!".format(asigra_postgresql_conf))
+                return
+
+            # listen_address = '*'
+            pg_conf_regex = re.compile('.*(#)\s{0,}?listen_addresses\s{0,}=\s{0,}([^\s]+)', re.M | re.S)
+            pg_conf_buf = []
+            rewrite = False
+
+            with open(asigra_postgresql_conf, "r") as f:
+                for line in f:
+                    if pg_conf_regex.search(line):
+                        def pg_re_replace(m):
+                            if m and len(m.groups()) > 1:
+                                return m.group(0)[m.end(1):m.start(2)] + "'*'" + m.group(0)[m.end(2):]
+
+                        line = pg_conf_regex.sub(pg_re_replace, line)
+                        rewrite = True
+                    pg_conf_buf.append(line)
+
+            if rewrite and pg_conf_buf:
+                with open(asigra_postgresql_conf, "w") as f:
+                    f.write("".join(pg_conf_buf))
+
+            if not os.path.exists(asigra_pg_hba_conf):
+                self.logger.error("{} doesn't exist!".format(asigra_pg_hba_conf))
+                return False
+
+            # allow local access
+            with open(asigra_pg_hba_conf, "a") as f:
+                f.write("host\tall\tall\t127.0.0.0/24\ttrust")
+
+        return True
+
+    # XXX
+    # We should probably write a postgresql middleware plugin
+    # We should also pass around a persistent DB handle
+    # XXX
+    async def asigra_database_exists(self):
+        exists = False
+        con = None
+
+        try:
+            con = pg.connect("dbname='template1' user='{}'".format(self.pg_user))
+
+        except Exception as e:
+            self.logger.error("Can't connect to template1: {}".format(e))
+            return exists
+
+        cur = con.cursor()
+        try:
+            cur.execute("SELECT COUNT(*) from pg_database where datname = 'dssystem'")
+
+        except Exception as e:
+            self.logger.error("Can't query dssystem!: {}".format(e))
+            return exists
+
+        exists = False
+        try:
+            exists = (cur.fetchone()[0] != 0)
+        except Exception:
+            exists = False
+
+        cur.close()
+        con.close()
+
+        return exists
+
+    async def asigra_database_create(self):
```

Comment:
Given all the blocking operations (database) this should not be a coroutine, but a normal python method.